apps/passport-server/src/services/generic-issuance/pipelines/LemonadePipeline.ts

```diff
@@ -1941,6 +1969,377 @@ export class LemonadePipeline implements BasePipeline {
+      // are no temporary states in which the checkers's permissions are somehow
+      // missing, e.g. due to a partial sync. This doesn't seem to be possible
+      // right now, but is something to be aware of if we make changes later.
+      // Denied offline check-ins are logged so that we can keep tabs on this.
+      const offlineCheckinsToSave: string[] = [];
+
+      for (const ticketId of ticketIds) {
+        const atom = await this.db.loadById(this.id, ticketId);
+        if (atom && atom.checkinDate === null) {
+          if (ticketMatchesCriteria(atom, checkerTicketCriteria)) {
+            offlineCheckinsToSave.push(ticketId);
+          } else {
+            logger(
+              `${LOG_TAG} User ${checkerEmail} tried to upload offline check-in for ticket ID ${ticketId} but is not permitted to do so.`
+            );
+          }
+        } else {
+          const manualTicket = this.getManualTicketById(ticketId);
+          if (manualTicket) {
+            if (ticketMatchesCriteria(manualTicket, checkerTicketCriteria)) {
+              const manualTicketCheckin = await this.checkinDB.getByTicketId(
+                this.id,
+                manualTicket.id
+              );
+              if (manualTicketCheckin === undefined) {
+                offlineCheckinsToSave.push(ticketId);
+              }
+            } else {
+              logger(
+                `${LOG_TAG} User ${checkerEmail} tried to upload offline check-in for ticket ID ${ticketId} but is not permitted to do so.`
+              );
+            }
+          }
+        }
+      }
+
+      span?.setAttribute(
+        "offline_checkins_added",
+        offlineCheckinsToSave.length
+      );
+
+      if (offlineCheckinsToSave.length > 0) {
+        await this.offlineCheckinDB.addOfflineCheckins(
+          this.id,
+          checkerEmail,
+          offlineCheckinsToSave,
+          new Date()
+        );
+      }
+    });
+  }
+
+  /**
+   * Performs a full check-in with the back-end for check-ins which were
+   * collected by an offline checker.
+   *
+   * This may involve checking in multiple tickets, from multiple different
+   * checkers. Due to the asynchronous nature of the check-in, we cannot report
+   * failure directly to the user, so we record a count of the failures.
+   */
+  private async processOfflineCheckins(): Promise<{
+    failedTicketIds: string[];
+    checkedInTicketIds: string[];
+    logs: PipelineLog[];
+  }> {
+    return traced(LOG_NAME, "processOfflineCheckins", async (span) => {
+      tracePipeline(this.definition);
+
+      const logs: PipelineLog[] = [];
+      const offlineCheckins =
+        await this.offlineCheckinDB.getOfflineCheckinsForPipeline(this.id);
+
+      const checkedInTicketIds = [];
+      const failedTicketIds = [];
+      span?.setAttribute("offline_checkin_count", offlineCheckins.length);
+
+      const ticketIdsGroupedByChecker = offlineCheckins.reduce(
```

Comment:
(`_.groupBy`)

Reply:
Same thing applies here. We're not turning offline checkins into a Record of emails to check-ins, we're turning it into a Record of emails to ticket IDs.